service/dataset.py

```diff
@@ -48,60 +74,165 @@ def read(dataset_request):
     except:
         # TODO: Temp fix for testing - Do better cleanup if secret missing
         # dataset["dataset_type"]["token"] = {'name':dataset_name,'value':42}
-        pass 
+        pass
+
+    return dataset
+
+def read(dataset_request):
+    """
+    Read from private dataset
+
+    :param info: The dataset to read and budget to use.
+    :type info: dict {"dataset_name": str, "budget":int}
+    :return: A dataset document that contains the type and info of the dataset
+    :rtype: dict{"dataset_type": str, dataset_key: dict}
+    """
+    return _read_helper(dataset_request, DATASETS)
+
+def readreleased(dataset_request):
+    """
+    Read from a released dataset, if authorized
     
+    Check request header for client_guid
+
+    :param info: The dataset to read and budget to use
+    :type info: dict {"dataset_name": str, "budget":int}
+    :return: A dataset document that contains the type and info of the dataset
+    :rtype: dict{"dataset_type": str, dataset_key: dict}
+    """
+    dataset = _read_helper(dataset_request, RELEASED_DATASETS)
+
+    client_guid = request.headers.get('client_guid')
+
+    # If client owns a released dataset, treat readreleased as
+    # private read
+    if client_guid in OWNERS:
+        if dataset_request["dataset_name"] in OWNERS[client_guid]:
+            return dataset
+
+    # NOTE: Should not happen
+    if 'authorized_users' not in dataset:
+        abort(411, "Released dataset must specify authorized users.")
+
+    if client_guid not in dataset['authorized_users']:
+        abort(404, "User not authorized to access this dataset.")
+
     # Check/Decrement the budget before returning dataset
-    adjusted_budget = dataset["budget"] - dataset_request["budget"]
-    if adjusted_budget >= 0.0:
-        dataset["budget"] = adjusted_budget
-    else:
-        abort(412, "Not enough budget for read. Remaining budget: {}".format(dataset_name))
+    # Note: We are guaranteed that client_guid in KNOWN_USERS
+    if client_guid not in KNOWN_USERS:
```

Comment:
From our chat I thought we landed on a simpler approach. No authorized users yet(that was a possible future thing)

For now we have USERS = {}
release_dataset(origi_dataset, budget, release_dataset_name) -> adds the owner to USERS_RELEASED[dataset_name][owner_id] += 1)
read_released_dataset(released_dataset_name) USERS_RELEASED[dataset_name][user_id] +=1 if USERS_RELEASED[dataset_name][user_id] > 1 charge budget

We can do a similar thing for authorized users if you want to add it where first we verify user_id in AUTHORIZED_USERS[released_dataset_name]  if not fail

Reply:
Budget is global so there is no need to block for new users since we are not separating budget between authorized users and the global pool.